File: Design Patterns/creational patterns/Factory/Notification Delivery System/typescript/creators.ts
import {
	EmailNotification,
	Notification,
	PushNotification,
	SMSNotification,
} from './products';

abstract class NotificationCreator implements NotificationCreator {
	public abstract createNotification(): Notification;

	public sendNotification(): void {
		const notification = this.createNotification();
		notification.send();
	}
}

class EmailNotificationCreator extends NotificationCreator {
	public createNotification(): Notification {
		return new EmailNotification();
	}
}

class SMSNotificationCreator extends NotificationCreator {
	public createNotification(): Notification {
		return new SMSNotification();
	}
}

class PushNotificationCreator extends NotificationCreator {
	public createNotification(): Notification {
		return new PushNotification();
	}
}

export {
	NotificationCreator,
	EmailNotificationCreator,
	SMSNotificationCreator,
	PushNotificationCreator,
};

File: Design Patterns/creational patterns/Factory/Notification Delivery System/java/creators.java
// Abstract Creator declaring the factory method
public abstract class NotificationCreator {

    // Factory method to be implemented by concrete creators
    public abstract Notification createNotification();


    // Business logic that uses the notification
    public void notifyUser(String message) {
        Notification notification = createNotification();
        notification.sendNotification(message);
    }
}

// Concrete Creator for Email Notifications
public class EmailNotificationCreator extends NotificationCreator {
    @Override
    public Notification createNotification() {
        return new EmailNotification();
    } 
}


// Concrete Creator for SMS Notifications
public class SMSNotificationCreator extends NotificationCreator {
    @Override
    public Notification createNotification() {
        return new SMSNotification();
    }
}
